File: AndroidSwiftUI/composeui/src/commonMain/kotlin/com/pureswift/swiftui/cupertino/CupertinoLayout.kt
@file:OptIn(io.github.alexzhirkevich.cupertino.ExperimentalCupertinoApi::class,
    io.github.alexzhirkevich.cupertino.adaptive.ExperimentalAdaptiveApi::class)
@file:Suppress("DEPRECATION")
package com.pureswift.swiftui

import androidx.compose.foundation.layout.*
import androidx.compose.runtime.*
import androidx.compose.ui.Modifier
import androidx.compose.ui.geometry.Rect
import androidx.compose.ui.graphics.Color
import androidx.compose.ui.unit.dp
import io.github.alexzhirkevich.LocalContentColor
import io.github.alexzhirkevich.cupertino.*
import io.github.alexzhirkevich.cupertino.adaptive.*
import io.github.alexzhirkevich.cupertino.section.CupertinoSectionDefaults
import io.github.alexzhirkevich.cupertino.theme.CupertinoTheme
import kotlinx.serialization.json.*

internal val LocalCupertinoScaffoldPadding = staticCompositionLocalOf { PaddingValues(0.dp) }
internal val LocalCupertinoScaffold = staticCompositionLocalOf { false }
private data class SegmentSlot(val selected: Int, val positions: List<TabPosition>)
private val LocalCupertinoSegment = staticCompositionLocalOf<SegmentSlot?> { null }

internal fun cupertinoInsets(node: ViewNode, key: String, default: WindowInsets): WindowInsets {
    if (node.string(key) == "none") return WindowInsets(0, 0, 0, 0)
    val obj = cupertinoObject(node, key) ?: return default
    return WindowInsets(left = (obj.double("left") ?: obj.double("start") ?: 0.0).toFloat().dp,
        top = (obj.double("top") ?: 0.0).toFloat().dp,
        right = (obj.double("right") ?: obj.double("end") ?: 0.0).toFloat().dp,
        bottom = (obj.double("bottom") ?: 0.0).toFloat().dp)
}

@Composable
internal fun RenderCupertinoLayout(node: ViewNode): Boolean {
    val name = node.string("name") ?: return false
    val props = Props(node.props)
    when (name) {
        "CupertinoSurface", "Surface", "AdaptiveSurface" -> {
            val shape = cupertinoShape(node)
            val contentColor = cupertinoColor(node, "contentColor") ?: LocalContentColor.current
            val color = cupertinoColor(node, "color") ?: CupertinoTheme.colorScheme.systemBackground
            val click = props.voidAction("onClick")
            val enabled = cupertinoEnabled(node)
            val content: @Composable () -> Unit = { cupertinoSlot(node, "content") }
            if (name == "AdaptiveSurface") AdaptiveSurface(modifier = node.composeModifiers(), shape = shape,
                color = cupertinoColor(node, "color") ?: Color.Unspecified,
                contentColor = cupertinoColor(node, "contentColor") ?: Color.Unspecified,
                shadowElevation = (node.double("shadowElevation") ?: 0.0).toFloat().dp, content = content)
            else if (click != null) CupertinoSurface(onClick = { if (enabled) click() }, modifier = node.composeModifiers(),
                enabled = enabled, shape = shape, color = color, contentColor = contentColor, border = cupertinoBorder(node), content = content)
            else CupertinoSurface(modifier = node.composeModifiers(), shape = shape, color = color,
                shadowElevation = (node.double("shadowElevation") ?: 0.0).toFloat().dp, contentColor = contentColor, content = content)
        }
        "CupertinoScaffold", "AdaptiveScaffold" -> {
            val content: @Composable (PaddingValues) -> Unit = { padding ->
                CompositionLocalProvider(LocalCupertinoScaffoldPadding provides padding) {
                    if (node.bool("applyContentPadding") ?: false) Box(Modifier.padding(padding)) { cupertinoSlot(node, "content") }
                    else cupertinoSlot(node, "content")
                }
            }
            val top: @Composable () -> Unit = { cupertinoSlot(node, "topBar") }
            val bottom: @Composable () -> Unit = { cupertinoSlot(node, "bottomBar") }
            val snackbar: @Composable () -> Unit = { cupertinoSlot(node, "snackbarHost") }
            val fab: @Composable () -> Unit = { cupertinoSlot(node, "floatingActionButton") }
            val fabPosition = if (node.string("floatingActionButtonPosition")?.lowercase() == "center") FabPosition.Center else FabPosition.End
            CompositionLocalProvider(LocalCupertinoScaffold provides true) {
                if (name == "CupertinoScaffold") CupertinoScaffold(modifier = node.composeModifiers(),
                    topBar = top, bottomBar = bottom, snackbarHost = snackbar, floatingActionButton = fab,
                    floatingActionButtonPosition = fabPosition,
                    containerColor = cupertinoColor(node, "containerColor") ?: CupertinoScaffoldDefaults.containerColor,
                    contentColor = cupertinoColor(node, "contentColor") ?: CupertinoScaffoldDefaults.contentColor,
                    contentWindowInsets = cupertinoInsets(node, "contentWindowInsets", CupertinoScaffoldDefaults.contentWindowInsets),
                    appBarsBlurAlpha = props.float("appBarsBlurAlpha") ?: CupertinoScaffoldDefaults.AppBarsBlurAlpha,
                    appBarsBlurRadius = props.float("appBarsBlurRadius")?.dp ?: CupertinoScaffoldDefaults.AppBarsBlurRadius,
                    hasNavigationTitle = node.bool("hasNavigationTitle") ?: false, content = content)
                else AdaptiveScaffold(modifier = node.composeModifiers(), topBar = top, bottomBar = bottom,
                    snackbarHost = snackbar, floatingActionButton = fab, floatingActionButtonPosition = fabPosition,
                    containerColor = cupertinoColor(node, "containerColor") ?: Color.Unspecified,
                    contentColor = cupertinoColor(node, "contentColor") ?: Color.Unspecified,
                    contentWindowInsets = cupertinoInsets(node, "contentWindowInsets", CupertinoScaffoldDefaults.contentWindowInsets), content = content)
            }
        }
        "CupertinoScaffoldPadding" -> Box(node.composeModifiers().padding(LocalCupertinoScaffoldPadding.current)) { cupertinoSlot(node, "content") }
        "CupertinoTopAppBar", "AdaptiveTopAppBar" -> {
            val title: @Composable () -> Unit = { cupertinoSlot(node, "title") }
            val navigation: @Composable () -> Unit = { cupertinoSlot(node, "navigationIcon") }
            val actions: @Composable RowScope.() -> Unit = { CompositionLocalProvider(LocalCupertinoRowScope provides this) { cupertinoSlot(node, "actions") } }
            val transparent = node.bool("isTransparent") ?: false
            val translucent = node.bool("isTranslucent") ?: LocalCupertinoScaffold.current
            val colors = CupertinoTopAppBarDefaults.topAppBarColors(
                containerColor = cupertinoColor(node, "containerColor") ?: CupertinoTheme.colorScheme.tertiarySystemBackground,
                scrolledContainerColor = cupertinoColor(node, "scrolledContainerColor") ?: Color.Transparent,
                navigationIconContentColor = cupertinoColor(node, "navigationIconContentColor") ?: CupertinoTheme.colorScheme.accent,
                titleContentColor = cupertinoColor(node, "titleContentColor") ?: CupertinoTheme.colorScheme.label,
                actionIconContentColor = cupertinoColor(node, "actionIconContentColor") ?: CupertinoTheme.colorScheme.accent)
            val divider: @Composable () -> Unit = {
                if (cupertinoHasSlot(node, "divider")) cupertinoSlot(node, "divider")
                else if (!transparent) CupertinoTopAppBarDefaults.divider()
            }
            if (name == "CupertinoTopAppBar") CupertinoTopAppBar(title, node.composeModifiers(), navigation, actions,
                windowInsets = cupertinoInsets(node, "windowInsets", CupertinoTopAppBarDefaults.windowInsets),
                isTransparent = transparent, isTranslucent = translucent, divider = divider, colors = colors)
            else AdaptiveTopAppBar(title, node.composeModifiers(), navigation, actions,
                windowInsets = cupertinoInsets(node, "windowInsets", CupertinoTopAppBarDefaults.windowInsets), adaptation = {
                    cupertino { this.colors = colors; isTransparent = transparent; isTranslucent = translucent; this.divider = { divider() } }
                    material { isCenterAligned = node.bool("isCenterAligned") ?: isCenterAligned }
                })
        }
        "CupertinoBottomAppBar" -> CupertinoBottomAppBar(modifier = node.composeModifiers(),
            isTranslucent = node.bool("isTranslucent") ?: true, isTransparent = node.bool("isTransparent") ?: false,
            containerColor = cupertinoColor(node, "containerColor") ?: CupertinoNavigationBarDefaults.containerColor,
            contentColor = cupertinoColor(node, "contentColor") ?: CupertinoTheme.colorScheme.accent,
            contentPadding = cupertinoPadding(node, "contentPadding", CupertinoSectionDefaults.PaddingValues),
            windowInsets = cupertinoInsets(node, "windowInsets", WindowInsets.navigationBars)) {
                CompositionLocalProvider(LocalCupertinoRowScope provides this) { cupertinoSlot(node, "content") }
            }
        "CupertinoNavigationBar", "AdaptiveNavigationBar" -> {
            val content: @Composable RowScope.() -> Unit = { CompositionLocalProvider(LocalCupertinoRowScope provides this) { cupertinoSlot(node, "content") } }
            if (name == "CupertinoNavigationBar") CupertinoNavigationBar(modifier = node.composeModifiers(),
                containerColor = cupertinoColor(node, "containerColor") ?: CupertinoNavigationBarDefaults.containerColor,
                windowInsets = cupertinoInsets(node, "windowInsets", WindowInsets.navigationBars),
                isTransparent = node.bool("isTransparent") ?: false, isTranslucent = node.bool("isTranslucent") ?: LocalCupertinoScaffold.current,
                divider = { if (cupertinoHasSlot(node, "divider")) cupertinoSlot(node, "divider") else CupertinoNavigationBarDefaults.divider() }, content = content)
            else AdaptiveNavigationBar(modifier = node.composeModifiers(),
                windowInsets = cupertinoInsets(node, "windowInsets", androidx.compose.material3.NavigationBarDefaults.windowInsets), content = content)
        }
        "RowScope.CupertinoNavigationBarItem", "CupertinoNavigationBarItem", "RowScope.AdaptiveNavigationBarItem", "AdaptiveNavigationBarItem" -> {
            val scope = LocalCupertinoRowScope.current
            if (scope == null) CupertinoText("$name requires a navigation bar RowScope", color = Color.Red)
            else with(scope) {
                val enabled = cupertinoEnabled(node)
                val click = props.voidAction("onClick")
                val action: () -> Unit = { if (enabled) click?.invoke() }
                val icon: @Composable () -> Unit = { cupertinoSlot(node, "icon") }
                val label: (@Composable () -> Unit)? = if (cupertinoHasSlot(node, "label")) ({ cupertinoSlot(node, "label") }) else null
                if (name.contains("Adaptive")) AdaptiveNavigationBarItem(node.bool("selected") ?: false, action, icon,
                    node.composeModifiers(), enabled, label, node.bool("alwaysShowLabel") ?: true)
                else CupertinoNavigationBarItem(node.bool("selected") ?: false, action, icon, node.composeModifiers(), enabled,
                    label, node.bool("alwaysShowLabel") ?: true, node.bool("pressIndicationEnabled") ?: false,
                    colors = CupertinoNavigationBarDefaults.itemColors(
                        selectedIconColor = cupertinoColor(node, "selectedIconColor") ?: CupertinoTheme.colorScheme.accent,
                        selectedTextColor = cupertinoColor(node, "selectedTextColor") ?: CupertinoTheme.colorScheme.accent,
                        unselectedIconColor = cupertinoColor(node, "unselectedIconColor") ?: CupertinoTheme.colorScheme.secondaryLabel,
                        unselectedTextColor = cupertinoColor(node, "unselectedTextColor") ?: CupertinoTheme.colorScheme.secondaryLabel,
                        disabledIconColor = cupertinoColor(node, "disabledIconColor") ?: CupertinoTheme.colorScheme.tertiaryLabel,
                        disabledTextColor = cupertinoColor(node, "disabledTextColor") ?: CupertinoTheme.colorScheme.tertiaryLabel))
            }
        }
        "CupertinoNavigationTitle" -> CupertinoNavigationTitle(modifier = node.composeModifiers(),
            maxFontScale = props.float("maxFontScale") ?: 1.1f,
            maxFontScaleDistance = props.float("maxFontScaleDistance")?.dp ?: 150.dp,
            paddingValues = cupertinoPadding(node, "paddingValues", CupertinoSectionDefaults.PaddingValues)) { cupertinoSlot(node, "content") }
        "CupertinoTopAppBarDefaults.divider" -> CupertinoTopAppBarDefaults.divider()
        "CupertinoNavigationBarDefaults.divider" -> CupertinoNavigationBarDefaults.divider()
        "CupertinoSegmentedControl" -> {
            val selected = (props.int("selectedTabIndex") ?: 0).coerceAtLeast(0)
            val defaults = CupertinoSegmentedControlDefaults.colors()
            val colors = CupertinoSegmentedControlDefaults.colors(
                containerColor = cupertinoColor(node, "containerColor") ?: defaults.containerColor,
                indicatorColor = cupertinoColor(node, "indicatorColor") ?: defaults.indicatorColor,
                contentColor = cupertinoColor(node, "contentColor") ?: defaults.contentColor,
                separatorColor = cupertinoColor(node, "separatorColor") ?: defaults.separatorColor)
            val shape = cupertinoShape(node, default = CupertinoSegmentedControlDefaults.shape)
            CupertinoSegmentedControl(selected, node.composeModifiers(), colors, shape,
                cupertinoPadding(node, "paddingValues", CupertinoSegmentedControlDefaults.PaddingValues), indicator = { positions ->
                    if (cupertinoHasSlot(node, "indicator")) CompositionLocalProvider(LocalCupertinoSegment provides SegmentSlot(selected, positions)) { cupertinoSlot(node, "indicator") }
                    else if (positions.isNotEmpty()) CupertinoSegmentedControlIndicator(selected.coerceIn(positions.indices), positions,
                        shape = shape, color = colors.indicatorColor, separatorColor = colors.separatorColor)
                }, tabs = { cupertinoSlot(node, "tabs") })
        }
        "CupertinoSegmentedControlTab" -> {
            val enabled = cupertinoEnabled(node)
            val action = props.voidAction("onClick")
            CupertinoSegmentedControlTab(onClick = { if (enabled) action?.invoke() }, isSelected = node.bool("isSelected") ?: false,
                modifier = node.composeModifiers()) { cupertinoSlot(node, "content") }
        }
        "CupertinoSegmentedControlIndicator" -> {
            val context = LocalCupertinoSegment.current
            if (context == null) CupertinoText("CupertinoSegmentedControlIndicator requires an indicator slot", color = Color.Red)
            else if (context.positions.isNotEmpty()) CupertinoSegmentedControlIndicator(
                (props.int("selectedTabIndex") ?: context.selected).coerceIn(context.positions.indices), context.positions,
                node.composeModifiers(), cupertinoShape(node, default = CupertinoTheme.shapes.small),
                cupertinoColor(node, "color") ?: CupertinoSegmentedControlDefaults.colors().indicatorColor,
                cupertinoColor(node, "separatorColor") ?: CupertinoTheme.colorScheme.separator)
        }
        "Modifier.haze" -> {
            val raw = node.props["area"] ?: node.props["areaJson"]
            val array = (raw as? JsonArray) ?: (raw as? JsonPrimitive)?.content?.let { runCatching { Json.parseToJsonElement(it) as? JsonArray }.getOrNull() }
            val areas = array.orEmpty().mapNotNull { value ->
                val area = value as? JsonObject ?: return@mapNotNull null
                val left = area.double("left")?.toFloat() ?: return@mapNotNull null
                val top = area.double("top")?.toFloat() ?: return@mapNotNull null
                val right = area.double("right")?.toFloat() ?: return@mapNotNull null
                val bottom = area.double("bottom")?.toFloat() ?: return@mapNotNull null
                Rect(left, top, right, bottom)
            }.toTypedArray()
            val background = cupertinoColor(node, "backgroundColor") ?: CupertinoTheme.colorScheme.systemBackground
            Box(node.composeModifiers().haze(*areas, backgroundColor = background,
                tint = cupertinoColor(node, "tint") ?: HazeDefaults.tint(background),
                blurRadius = props.float("blurRadius")?.dp ?: HazeDefaults.blurRadius)) { cupertinoSlot(node, "content") }
        }
        "TabRowDefaults.Modifier.tabIndicatorOffset" -> {
            val context = LocalCupertinoSegment.current
            val position = context?.positions?.getOrNull(props.int("selectedTabIndex") ?: context.selected)
            if (position == null) CupertinoText("tabIndicatorOffset requires a segmented-control indicator slot", color = Color.Red)
            else with(TabRowDefaults) { Box(node.composeModifiers().tabIndicatorOffset(position)) { cupertinoSlot(node, "content") } }
        }
        "Modifier.cupertinoPickerIndicator" -> {
            val state = LocalCupertinoPickerState.current
            if (state == null) CupertinoText("cupertinoPickerIndicator requires the measured picker state scope", color = Color.Red)
            else {
                val indicatorNode = node.copy(props = JsonObject(node.props + buildMap {
                    node.props["shape"]?.let { put("indicatorShape", it) }
                    node.props["paddingValues"]?.let { put("indicatorPadding", it) }
                }))
                val indicator = cupertinoWheelIndicator(indicatorNode)
                Box(node.composeModifiers().cupertinoPickerIndicator(state, indicator)) { cupertinoSlot(node, "content") }
            }
        }
        else -> return false
    }
    return true
}

File: AndroidSwiftUI/SwiftUICore/Sources/SwiftUICore/Cupertino/CupertinoLayout.swift
public enum CupertinoColorRole: String, CaseIterable, Sendable {
    case accent
    case label
    case secondaryLabel
    case tertiaryLabel
    case quaternaryLabel
    case systemFill
    case secondarySystemFill
    case tertiarySystemFill
    case quaternarySystemFill
    case placeholderText
    case separator
    case opaqueSeparator
    case link
    case systemGroupedBackground
    case secondarySystemGroupedBackground
    case tertiarySystemGroupedBackground
    case systemBackground
    case secondarySystemBackground
    case tertiarySystemBackground
}
public enum CupertinoShapeSize: String, CaseIterable, Sendable { case extraSmall, small, medium, large, extraLarge }
public struct CupertinoThemeOptions: Sendable {
    public var target: CupertinoTarget
    public var isDark: Bool?
    public var colors: [CupertinoColorRole: Color]
    public var shapes: [CupertinoShapeSize: Double]
    public var fontFamily: String?
    public init(target: CupertinoTarget = .cupertino, isDark: Bool? = nil,
                colors: [CupertinoColorRole: Color] = [:], shapes: [CupertinoShapeSize: Double] = [:], fontFamily: String? = nil) {
        precondition(shapes.values.allSatisfy { $0.isFinite && $0 >= 0 })
        self.target = target; self.isDark = isDark; self.colors = colors; self.shapes = shapes; self.fontFamily = fontFamily
    }
    var props: [String: PropValue] {
        var p: [String: PropValue] = ["target": .string(target.rawValue)]
        if let isDark { p["isDark"] = .bool(isDark) }
        if let fontFamily { p["fontFamily"] = .string(fontFamily) }
        for (key, color) in colors { p[key.rawValue] = .color(color) }
        for (key, radius) in shapes { p[key.rawValue] = .double(radius) }
        return p
    }
}

public struct CupertinoTheme: View {
    let component: CupertinoComponent
    public init<Content: View>(_ options: CupertinoThemeOptions = .init(), @ViewBuilder content: () -> Content) {
        component = CupertinoComponent("CupertinoTheme", props: options.props, slots: [CupertinoSlot("content", content: content)])
    }
    public var body: some View { component }
}

public struct AdaptiveTheme: View {
    let component: CupertinoComponent
    public init<Content: View>(_ options: CupertinoThemeOptions = .init(), @ViewBuilder content: () -> Content) {
        component = CupertinoComponent("AdaptiveTheme", props: options.props, slots: [CupertinoSlot("content", content: content)])
    }
    public var body: some View { component }
}

public struct AdaptiveWidget: View {
    let component: CupertinoComponent
    public init<C: View, M: View>(@ViewBuilder cupertino: () -> C, @ViewBuilder material: () -> M) {
        component = CupertinoComponent("AdaptiveWidget", slots: [CupertinoSlot("cupertino", content: cupertino), CupertinoSlot("material", content: material)])
    }
    public var body: some View { component }
}

/// Common typed options shared by surfaces, bars and scaffolds; named slots carry
/// every upstream content position without flattening title/actions into a label.
public struct CupertinoContainerOptions: Sendable {
    public var enabled: Bool = true
    public var color: Color?
    public var contentColor: Color?
    public var containerColor: Color?
    public var shadowElevation: Double = 0
    public var isTransparent: Bool = false
    public var isTranslucent: Bool = false
    public var hasNavigationTitle: Bool = false
    public var applyContentPadding: Bool = true
    public init() {}
    var props: [String: PropValue] {
        precondition(shadowElevation.isFinite && shadowElevation >= 0)
        var p: [String: PropValue] = ["enabled": .bool(enabled), "shadowElevation": .double(shadowElevation),
            "isTransparent": .bool(isTransparent), "isTranslucent": .bool(isTranslucent), "hasNavigationTitle": .bool(hasNavigationTitle),
            "applyContentPadding": .bool(applyContentPadding)]
        if let color { p["color"] = .color(color) }
        if let contentColor { p["contentColor"] = .color(contentColor) }
        if let containerColor { p["containerColor"] = .color(containerColor) }
        return p
    }
}

public struct CupertinoSurface: View {
    let component: CupertinoComponent
    public init(options: CupertinoContainerOptions = .init(), onClick: (() -> Void)? = nil, slots: [CupertinoSlot]) {
        var actions: [String: ComposableAction] = [:]
        if let onClick { actions["onClick"] = .void { if options.enabled { onClick() } } }
        component = CupertinoComponent("CupertinoSurface", props: options.props, actions: actions, slots: slots)
    }
    public init<Content: View>(options: CupertinoContainerOptions = .init(), @ViewBuilder content: () -> Content) {
        self.init(options: options, slots: [CupertinoSlot("content", content: content)])
    }
    public var body: some View { component }
}

public struct AdaptiveSurface: View {
    let component: CupertinoComponent
    public init(options: CupertinoContainerOptions = .init(), onClick: (() -> Void)? = nil, slots: [CupertinoSlot]) {
        var actions: [String: ComposableAction] = [:]
        if let onClick { actions["onClick"] = .void { if options.enabled { onClick() } } }
        component = CupertinoComponent("AdaptiveSurface", props: options.props, actions: actions, slots: slots)
    }
    public init<Content: View>(options: CupertinoContainerOptions = .init(), @ViewBuilder content: () -> Content) {
        self.init(options: options, slots: [CupertinoSlot("content", content: content)])
    }
    public var body: some View { component }
}

public struct CupertinoScaffold: View {
    let component: CupertinoComponent
    public init(options: CupertinoContainerOptions = .init(), onClick: (() -> Void)? = nil, slots: [CupertinoSlot]) {
        var actions: [String: ComposableAction] = [:]
        if let onClick { actions["onClick"] = .void { if options.enabled { onClick() } } }
        component = CupertinoComponent("CupertinoScaffold", props: options.props, actions: actions, slots: slots)
    }
    public init<Content: View>(options: CupertinoContainerOptions = .init(), @ViewBuilder content: () -> Content) {
        self.init(options: options, slots: [CupertinoSlot("content", content: content)])
    }
    public var body: some View { component }
}

public struct AdaptiveScaffold: View {
    let component: CupertinoComponent
    public init(options: CupertinoContainerOptions = .init(), onClick: (() -> Void)? = nil, slots: [CupertinoSlot]) {
        var actions: [String: ComposableAction] = [:]
        if let onClick { actions["onClick"] = .void { if options.enabled { onClick() } } }
        component = CupertinoComponent("AdaptiveScaffold", props: options.props, actions: actions, slots: slots)
    }
    public init<Content: View>(options: CupertinoContainerOptions = .init(), @ViewBuilder content: () -> Content) {
        self.init(options: options, slots: [CupertinoSlot("content", content: content)])
    }
    public var body: some View { component }
}

public struct CupertinoTopAppBar: View {
    let component: CupertinoComponent
    public init(options: CupertinoContainerOptions = .init(), onClick: (() -> Void)? = nil, slots: [CupertinoSlot]) {
        var actions: [String: ComposableAction] = [:]
        if let onClick { actions["onClick"] = .void { if options.enabled { onClick() } } }
        component = CupertinoComponent("CupertinoTopAppBar", props: options.props, actions: actions, slots: slots)
    }
    public init<Content: View>(options: CupertinoContainerOptions = .init(), @ViewBuilder content: () -> Content) {
        self.init(options: options, slots: [CupertinoSlot("content", content: content)])
    }
    public var body: some View { component }
}

public struct AdaptiveTopAppBar: View {
    let component: CupertinoComponent
    public init(options: CupertinoContainerOptions = .init(), onClick: (() -> Void)? = nil, slots: [CupertinoSlot]) {
        var actions: [String: ComposableAction] = [:]
        if let onClick { actions["onClick"] = .void { if options.enabled { onClick() } } }
        component = CupertinoComponent("AdaptiveTopAppBar", props: options.props, actions: actions, slots: slots)
    }
    public init<Content: View>(options: CupertinoContainerOptions = .init(), @ViewBuilder content: () -> Content) {
        self.init(options: options, slots: [CupertinoSlot("content", content: content)])
    }
    public var body: some View { component }
}

public struct CupertinoBottomAppBar: View {
    let component: CupertinoComponent
    public init(options: CupertinoContainerOptions = .init(), onClick: (() -> Void)? = nil, slots: [CupertinoSlot]) {
        var actions: [String: ComposableAction] = [:]
        if let onClick { actions["onClick"] = .void { if options.enabled { onClick() } } }
        component = CupertinoComponent("CupertinoBottomAppBar", props: options.props, actions: actions, slots: slots)
    }
    public init<Content: View>(options: CupertinoContainerOptions = .init(), @ViewBuilder content: () -> Content) {
        self.init(options: options, slots: [CupertinoSlot("content", content: content)])
    }
    public var body: some View { component }
}

public struct CupertinoNavigationBar: View {
    let component: CupertinoComponent
    public init(options: CupertinoContainerOptions = .init(), onClick: (() -> Void)? = nil, slots: [CupertinoSlot]) {
        var actions: [String: ComposableAction] = [:]
        if let onClick { actions["onClick"] = .void { if options.enabled { onClick() } } }
        component = CupertinoComponent("CupertinoNavigationBar", props: options.props, actions: actions, slots: slots)
    }
    public init<Content: View>(options: CupertinoContainerOptions = .init(), @ViewBuilder content: () -> Content) {
        self.init(options: options, slots: [CupertinoSlot("content", content: content)])
    }
    public var body: some View { component }
}

public struct AdaptiveNavigationBar: View {
    let component: CupertinoComponent
    public init(options: CupertinoContainerOptions = .init(), onClick: (() -> Void)? = nil, slots: [CupertinoSlot]) {
        var actions: [String: ComposableAction] = [:]
        if let onClick { actions["onClick"] = .void { if options.enabled { onClick() } } }
        component = CupertinoComponent("AdaptiveNavigationBar", props: options.props, actions: actions, slots: slots)
    }
    public init<Content: View>(options: CupertinoContainerOptions = .init(), @ViewBuilder content: () -> Content) {
        self.init(options: options, slots: [CupertinoSlot("content", content: content)])
    }
    public var body: some View { component }
}

public struct CupertinoNavigationTitle: View {
    let component: CupertinoComponent
    public init(options: CupertinoContainerOptions = .init(), onClick: (() -> Void)? = nil, slots: [CupertinoSlot]) {
        var actions: [String: ComposableAction] = [:]
        if let onClick { actions["onClick"] = .void { if options.enabled { onClick() } } }
        component = CupertinoComponent("CupertinoNavigationTitle", props: options.props, actions: actions, slots: slots)
    }
    public init<Content: View>(options: CupertinoContainerOptions = .init(), @ViewBuilder content: () -> Content) {
        self.init(options: options, slots: [CupertinoSlot("content", content: content)])
    }
    public var body: some View { component }
}

public struct CupertinoBottomSheetContent: View {
    let component: CupertinoComponent
    public init(options: CupertinoContainerOptions = .init(), onClick: (() -> Void)? = nil, slots: [CupertinoSlot]) {
        var actions: [String: ComposableAction] = [:]
        if let onClick { actions["onClick"] = .void { if options.enabled { onClick() } } }
        component = CupertinoComponent("CupertinoBottomSheetContent", props: options.props, actions: actions, slots: slots)
    }
    public init<Content: View>(options: CupertinoContainerOptions = .init(), @ViewBuilder content: () -> Content) {
        self.init(options: options, slots: [CupertinoSlot("content", content: content)])
    }
    public var body: some View { component }
}

public struct CupertinoNavigationBarItem: View {
    let component: CupertinoComponent
    public init(selected: Bool, enabled: Bool = true, onClick: @escaping () -> Void, slots: [CupertinoSlot]) {
        component = CupertinoComponent("RowScope.CupertinoNavigationBarItem", props: ["selected": .bool(selected), "enabled": .bool(enabled)],
            actions: ["onClick": .void { if enabled { onClick() } }], slots: slots)
    }
    public var body: some View { component }
}

public struct AdaptiveNavigationBarItem: View {
    let component: CupertinoComponent
    public init(selected: Bool, enabled: Bool = true, onClick: @escaping () -> Void, slots: [CupertinoSlot]) {
        component = CupertinoComponent("RowScope.AdaptiveNavigationBarItem", props: ["selected": .bool(selected), "enabled": .bool(enabled)],
            actions: ["onClick": .void { if enabled { onClick() } }], slots: slots)
    }
    public var body: some View { component }
}

public struct CupertinoSegmentedControl: View {
    let component: CupertinoComponent
    public init<Tabs: View>(selectedTabIndex: Int, slots: [CupertinoSlot] = [], @ViewBuilder tabs: () -> Tabs) {
        precondition(selectedTabIndex >= 0)
        component = CupertinoComponent("CupertinoSegmentedControl", props: ["selectedTabIndex": .int(selectedTabIndex)],
            slots: [CupertinoSlot("tabs", content: tabs)] + slots)
    }
    public var body: some View { component }
}
public struct CupertinoSegmentedControlTab: View {
    let component: CupertinoComponent
    public init<Content: View>(isSelected: Bool, onClick: @escaping () -> Void, @ViewBuilder content: () -> Content) {
        component = CupertinoComponent("CupertinoSegmentedControlTab", props: ["isSelected": .bool(isSelected)],
            actions: ["onClick": .void(onClick)], slots: [CupertinoSlot("content", content: content)])
    }
    public var body: some View { component }
}
public struct CupertinoSegmentedControlIndicator: View {
    let component: CupertinoComponent
    public init(selectedTabIndex: Int, color: Color? = nil) {
        precondition(selectedTabIndex >= 0)
        var p: [String: PropValue] = ["selectedTabIndex": .int(selectedTabIndex)]
        if let color { p["color"] = .color(color) }
        component = CupertinoComponent("CupertinoSegmentedControlIndicator", props: p)
    }
    public var body: some View { component }
}
